lambci/sync-threads
# Bugfix and performance improvements

Hi @mhart,

I found a bug in the library and got a bit carried away when fixing it, this PR contains the following (each in their own commit, so can be reviewed separately if needed)

- Tests for common features
- A bugfix (moving the SharedArrayBuffer inside the function call, because if it was not done, the function would always return the result of the first call)
- Support for multiple arguments in async functions (with a test)
- New argument for timeouts (with a test)
- Use a single worker for a given async function (big performance improvement, see below)
- Add a benchmark

Reusing the worker across calls is a significant performance improvement as you can see:

```
Load time
┌─────────┬────────────────────────┬───────────┬────────────────────┬──────────┬─────────┐
│ (index) │       Task Name        │  ops/sec  │ Average Time (ns)  │  Margin  │ Samples │
├─────────┼────────────────────────┼───────────┼────────────────────┼──────────┼─────────┤
│    0    │        'native'        │ '818 234' │ 1222.1439571458939 │ '±3.69%' │  81824  │
│    1    │     'sync-threads'     │ '927 568' │ 1078.087884336715  │ '±0.94%' │  92757  │
│    2    │ 'sync-threads (1.0.1)' │ '902 893' │ 1107.5497360136544 │ '±2.00%' │  90290  │
│    3    │       'synckit'        │ '958 314' │ 1043.4987350162614 │ '±2.21%' │  95832  │
└─────────┴────────────────────────┴───────────┴────────────────────┴──────────┴─────────┘
Execution time
┌─────────┬────────────────────────┬───────────┬────────────────────┬───────────┬─────────┐
│ (index) │       Task Name        │  ops/sec  │ Average Time (ns)  │  Margin   │ Samples │
├─────────┼────────────────────────┼───────────┼────────────────────┼───────────┼─────────┤
│    0    │        'native'        │ '127 639' │ 7834.536944837339  │ '±3.20%'  │  12764  │
│    1    │     'sync-threads'     │ '10 695'  │  93495.3008379255  │ '±7.74%'  │  1071   │
│    2    │ 'sync-threads (1.0.1)' │   '881'   │ 1134772.9650776037 │ '±49.95%' │   89    │
│    3    │       'synckit'        │  '1 789'  │ 558809.5973989817  │ '±22.36%' │   179   │
└─────────┴────────────────────────┴───────────┴────────────────────┴───────────┴─────────┘
Total time
┌─────────┬────────────────────────┬──────────┬────────────────────┬───────────┬─────────┐
│ (index) │       Task Name        │ ops/sec  │ Average Time (ns)  │  Margin   │ Samples │
├─────────┼────────────────────────┼──────────┼────────────────────┼───────────┼─────────┤
│    0    │        'native'        │ '73 668' │ 13574.320538260483 │ '±10.58%' │  7367   │
│    1    │     'sync-threads'     │ '2 169'  │ 460920.5601951494  │ '±7.17%'  │   217   │
│    2    │ 'sync-threads (1.0.1)' │ '1 228'  │ 813877.0010413193  │ '±58.56%' │   123   │
│    3    │       'synckit'        │  '761'   │ 1313029.3481143904 │ '±41.65%' │   81    │
└─────────┴────────────────────────┴──────────┴────────────────────┴───────────┴─────────┘

```

These tests were run today on my own MacBook Air m2.
I did not push the test for `sync-threads (1.0.1)` as it's just to illustrate the point.
